GcExtensions.md

```diff
@@ -0,0 +1,353 @@
+
+    ./autogen.sh
+    ./configure --with-gc=julia --with-julia=/path/to/julia/usr
+    make
+
+## Callbacks
+
+In order to allow foreign code to have access to necessary functionality
+in the garbage collector, we allow foreign code to register callbacks for
+certain GC events. We provide for six types of callbacks:
+
+* `pre_gc` (beginning of garbage collection).
+* `post_gc` (end of garbage collection).
+* `root_scanner` (before scanning GC roots).
+* `task_scanner` (when scanning Julia tasks).
+* `notify_external_alloc` (when an external object is allocated).
+* `notify_external_free` (when an external object is freed).
+
+With external objects, we refer to what in the current implementation are
+called `bigval_t` objects. These are allocated using the system's memory
+allocator rather than using Julia's external allocator. In order to not
+expose the implementation detail, we talk about "internal" and "external"
+objects rather than objects that are allocated as part of Julia's object
+pool or through system routines, respectively.
+
+For each type of callback, there is a corresponding function pointer type.
+Registering and deregistering callbacks occurs via macros.
+
+```
+typedef void (*jl_gc_cb_pre_gc_t)(int full);
+typedef void (*jl_gc_cb_post_gc_t)(int full);
+typedef void (*jl_gc_cb_root_scanner_t)(int full);
+typedef void (*jl_gc_cb_task_scanner_t)(jl_task_t *task, int full);
+typedef void (*jl_gc_cb_external_alloc_t)(void *addr, size_t size);
+typedef void (*jl_gc_cb_external_free_t)(void *addr);
+
+#define jl_gc_register_callback(cb, func)
+#define jl_gc_deregister_callback(cb, func)
+```
+
+The `jl_gc_register_callback()` and `jl_gc_deregister_callback()` macros
+are called with the appropriate index and a function pointer of the
+matching type. Each function can be registered at most once; duplicate
+registrations will be discarded.
+
+*Performance impact:* The callback implementation is designed to incur
+negligible overhead if no callbacks are used and no more overhead than
+necessary to invoke the callbacks. The callbacks are all kept in linked
+lists; if no callbacks are registered, all that is done is testing a
+static variable for being null and to branch if it is. As branch
+behavior should always be the same, only a few clock cycles are used, as
+long as the variable is in the cache and the branch target in the BTB.
+
+## Additional GC roots and hooking into the GC process
+
+We provide three callbacks that are called at the beginning of a GC
+(`pre_gc`), the beginning of the mark phase (`root_scanner`),
+and the end of the GC (`post_gc`).
+
+As these callbacks are tested and called only once per collection, overhead
+should be negligible.
+
+Additional roots can be marked from the root scanner callback by
+calling the `jl_gc_mark_queue_obj()` function, which takes a pointer to
+the current thread's thread-local storage a pointer to the object as its
+parameters.
```

Comment:
So this will mean that the scan of foreign object has to be done with breath first order, which will take much more space. If that's OK for the use case, it's fine. Otherwise, a much more integrated (and necessarily much less stable) API will be required.

Reply:
One option I have considered for this situation is to expose an interface to object arrays, which would handle that, but only works for object layouts where pointers are stored in consecutive locations.